pkg/selectors/selectors.go

```diff
@@ -75,6 +75,17 @@ type KernelLPMTrieStringPostfix struct {
 	data      [StringPostfixMaxLength]byte
 }
 
+type KernelSelectorMaps struct {
+	// stringMaps are used to populate string and char buf matches
+	stringMaps StringMapLists
+	// stringPrefixMaps are used to populate string and char buf prefix matches
+	stringPrefixMaps []map[KernelLPMTrieStringPrefix]struct{}
+	// stringPostfixMaps are used to populate string and char buf postfix matches
+	stringPostfixMaps []map[KernelLPMTrieStringPostfix]struct{}
+}
+
+var gMaps *KernelSelectorMaps
```

Comment:
Let's not use a global variable.

How about:
```diff
diff --git a/pkg/selectors/kernel.go b/pkg/selectors/kernel.go
index e3fa396a9..dc7ce8077 100644
--- a/pkg/selectors/kernel.go
+++ b/pkg/selectors/kernel.go
@@ -1258,7 +1258,7 @@ func parseSelector(
 //
 // For some examples, see kernel_test.go
 func InitKernelSelectors(selectors []v1alpha1.KProbeSelector, args []v1alpha1.KProbeArg, actionArgTable *idtable.Table) ([4096]byte, error) {
-	kernelSelectors, err := InitKernelSelectorState(selectors, args, actionArgTable, nil, true)
+	kernelSelectors, err := InitKernelSelectorState(selectors, args, actionArgTable, nil, nil)
 	if err != nil {
 		return [4096]byte{}, err
 	}
@@ -1266,8 +1266,8 @@ func InitKernelSelectors(selectors []v1alpha1.KProbeSelector, args []v1alpha1.KP
 }
 
 func InitKernelSelectorState(selectors []v1alpha1.KProbeSelector, args []v1alpha1.KProbeArg,
-	actionArgTable *idtable.Table, listReader ValueReader, newMaps bool) (*KernelSelectorState, error) {
-	kernelSelectors := NewKernelSelectorState(listReader, newMaps)
+	actionArgTable *idtable.Table, listReader ValueReader, maps *KernelSelectorMaps) (*KernelSelectorState, error) {
+	kernelSelectors := NewKernelSelectorState(listReader, maps)
 
 	WriteSelectorUint32(kernelSelectors, uint32(len(selectors)))
 	soff := make([]uint32, len(selectors))
diff --git a/pkg/selectors/selectors.go b/pkg/selectors/selectors.go
index 423549b73..6d6de5c3b 100644
--- a/pkg/selectors/selectors.go
+++ b/pkg/selectors/selectors.go
@@ -84,8 +84,6 @@ type KernelSelectorMaps struct {
 	stringPostfixMaps []map[KernelLPMTrieStringPostfix]struct{}
 }
 
-var gMaps *KernelSelectorMaps
-
 type KernelSelectorState struct {
 	off uint32     // offset into encoding
 	e   [4096]byte // kernel encoding of selectors
@@ -107,15 +105,15 @@ type KernelSelectorState struct {
 	maps *KernelSelectorMaps
 }
 
-func NewKernelSelectorState(listReader ValueReader, newMaps bool) *KernelSelectorState {
-	if newMaps {
-		gMaps = &KernelSelectorMaps{}
+func NewKernelSelectorState(listReader ValueReader, maps *KernelSelectorMaps) *KernelSelectorState {
+	if maps == nil {
+		maps = &KernelSelectorMaps{}
 	}
 	return &KernelSelectorState{
 		matchBinaries: make(map[int]*MatchBinariesMappings),
 		newBinVals:    make(map[uint32]string),
 		listReader:    listReader,
-		maps:          gMaps,
+		maps:          maps,
 	}
 }
 
diff --git a/pkg/sensors/tracing/generickprobe.go b/pkg/sensors/tracing/generickprobe.go
index b7d41f1dc..22eef9d7c 100644
--- a/pkg/sensors/tracing/generickprobe.go
+++ b/pkg/sensors/tracing/generickprobe.go
@@ -527,8 +527,14 @@ func createGenericKprobeSensor(
 		// Syscall flag might be changed in list definition
 		kprobes[i].Syscall = syscall
 
+		selMaps := &selectors.KernelSelectorMaps{}
 		for idx := range syms {
-			out, err := addKprobe(i, syms[idx], &kprobes[i], &in)
+			// if we use kprobe-multi, we need to share the maps.
+			// In all other cases, we create new ones.
+			if !useMulti && idx > 0 {
+				selMaps = &selectors.KernelSelectorMaps{}
+			}
+			out, err := addKprobe(i, syms[idx], &kprobes[i], &in, selMaps)
 			if err != nil {
 				return nil, err
 			}
@@ -581,7 +587,7 @@ func createGenericKprobeSensor(
 // addKprobe will, amongst other things, create a generic kprobe entry and add
 // it to the genericKprobeTable. The caller should make sure that this entry is
 // properly removed on kprobe unload.
-func addKprobe(index int, funcName string, f *v1alpha1.KProbeSpec, in *addKprobeIn) (out *addKprobeOut, err error) {
+func addKprobe(index int, funcName string, f *v1alpha1.KProbeSpec, in *addKprobeIn, selMaps *selectors.KernelSelectorMaps) (out *addKprobeOut, err error) {
 	var argSigPrinters []argPrinters
 	var argReturnPrinters []argPrinters
 	var setRetprobe bool
@@ -747,10 +753,7 @@ func addKprobe(index int, funcName string, f *v1alpha1.KProbeSpec, in *addKprobe
 	}
 
 	// Parse Filters into kernel filter logic
-	// If we use kprobe-multi and this is the first kprobe, we initialize new selector maps.
-	// In all the other cases we use shared maps for all attach points. In the case of
-	// non-kprobe-multi each kprobe has it's own maps.
-	kprobeEntry.loadArgs.selectors, err = selectors.InitKernelSelectorState(f.Selectors, f.Args, &kprobeEntry.actionArgs, nil, !in.useMulti || (in.useMulti && index == 0))
+	kprobeEntry.loadArgs.selectors, err = selectors.InitKernelSelectorState(f.Selectors, f.Args, &kprobeEntry.actionArgs, nil, selMaps)
 	if err != nil {
 		return nil, err
 	}
diff --git a/pkg/sensors/tracing/generictracepoint.go b/pkg/sensors/tracing/generictracepoint.go
index 46dee7953..5af8b3b96 100644
--- a/pkg/sensors/tracing/generictracepoint.go
+++ b/pkg/sensors/tracing/generictracepoint.go
@@ -508,7 +508,7 @@ func (tp *genericTracepoint) InitKernelSelectors(lists []v1alpha1.ListSpec) erro
 		}
 	}
 
-	selectors, err := selectors.InitKernelSelectorState(selSelectors, selArgs, &tp.actionArgs, &listReader{lists}, true)
+	selectors, err := selectors.InitKernelSelectorState(selSelectors, selArgs, &tp.actionArgs, &listReader{lists}, nil)
 	if err != nil {
 		return err
 	}
diff --git a/pkg/sensors/tracing/genericuprobe.go b/pkg/sensors/tracing/genericuprobe.go
index 6c6c5b158..1264a7af3 100644
--- a/pkg/sensors/tracing/genericuprobe.go
+++ b/pkg/sensors/tracing/genericuprobe.go
@@ -200,7 +200,7 @@ func createGenericUprobeSensor(
 		}
 
 		// Parse Filters into kernel filter logic
-		uprobeSelectorState, err := selectors.InitKernelSelectorState(spec.Selectors, args, nil, nil, true)
+		uprobeSelectorState, err := selectors.InitKernelSelectorState(spec.Selectors, args, nil, nil, nil)
 		if err != nil {
 			return nil, err
 		}
```

Reply:
Updated the PR to not use a global variable.